pkg/querier/parquet_queryable.go

```diff
@@ -0,0 +1,1031 @@
+		sts.AddParquetReaderCacheMiss(readerCacheMiss)
+	}()
+	for i, block := range blocks {
+		id := block.ID.String()
+		name := path.Join(uname, id, parquetFileName)
+
+		// Try to get from cache.
+		if reader := q.readerCache.Get(name); reader != nil {
+			output[i] = reader
+			continue
+		}
+		readerCacheMiss++
+
+		eg.Go(func() error {
+			attr, err := q.bucket.Attributes(ctx, name)
+			if err != nil {
+				if q.bucket.IsObjNotFoundErr(err) {
+					return nil
+				}
+				return err
+			}
+			t := time.Now()
+			r := &bReadAt{path: name, obj: q.bucket, m: q.metrics}
+			pr, err := mimir_storage.NewParquetReader(r.CreateReadAtWithContext, attr.Size, q.asyncRead, q.cacheMetrics, q.dictionaryCacheSize, q.metrics.pagesSkippedPageIndex, q.metrics.projectionPushdown)
+			if err != nil {
+				return err
+			}
+			mtx.Lock()
+			output[i] = pr
+			mtx.Unlock()
+			q.readerCache.Set(name, pr)
+			q.metrics.parquetFileOpenDuration.Observe(time.Since(t).Seconds())
+			return err
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+	return output, nil
+}
+
+func parquetValuesToStrings(values []parquet.Value) []string {
+	s := make(map[string]struct{}, len(values))
+	for _, value := range values {
+		s[value.String()] = struct{}{}
+	}
+	result := make([]string, 0, len(s))
+	for k := range s {
+		result = append(result, k)
+	}
+	sort.Strings(result)
+	return result
+}
+
+type RemoveHashLabelSeriesSet struct {
+	s       storage.SeriesSet
+	builder *labels.Builder
+}
+
+func newRemoveHashLabelSeriesSet(s storage.SeriesSet) *RemoveHashLabelSeriesSet {
+	return &RemoveHashLabelSeriesSet{s: s, builder: labels.NewBuilder(labels.EmptyLabels())}
+}
+
+func (ss *RemoveHashLabelSeriesSet) Next() bool {
+	return ss.s.Next()
+}
+
+// At returns full series. Returned series should be iterable even after Next is called.
+func (ss *RemoveHashLabelSeriesSet) At() storage.Series {
+	series := ss.s.At()
+	ss.builder.Reset(series.Labels())
+
+	return &storage.SeriesEntry{
+		Lset:             ss.builder.Del(seriesHash).Labels(),
+		SampleIteratorFn: series.Iterator,
+	}
+}
+
+// The error that iteration as failed with.
+// When an error occurs, set cannot continue to iterate.
+func (ss *RemoveHashLabelSeriesSet) Err() error {
+	return ss.s.Err()
+}
+
+// A collection of warnings for the whole set.
+// Warnings could be return even iteration has not failed with error.
+func (ss *RemoveHashLabelSeriesSet) Warnings() annotations.Annotations {
+	return ss.s.Warnings()
+}
+
+type ParquetRowsSeriesSet struct {
+	series []Series
+	curr   Series
+	err    error
+}
+
+type Series struct {
+	lbls       labels.Labels
+	chks       []chunks.Meta
+	mint, maxt int64
+}
+
+func (s Series) Labels() labels.Labels {
+	return s.lbls
+}
+
+func (s Series) Iterator(iterator chunkenc.Iterator) chunkenc.Iterator {
+	iters := make([]chunkenc.Iterator, 0, len(s.chks))
+	for _, chk := range s.chks {
+		iters = append(iters, chk.Chunk.Iterator(nil))
+	}
+	// TODO: honour mint and maxt
```

Comment:
Bringing visibility

Reply:
Addressed as part of https://github.com/grafana/mimir/pull/11270/commits/41ffa1526e807e44544d32a091c5c24c012c542d